api/src/main/java/org/pmiops/workbench/iam/IamServiceImpl.java

```diff
@@ -24,40 +27,64 @@
   private final CloudIamClient cloudIamClient;
   private final CloudResourceManagerService cloudResourceManagerService;
-  private final Provider<GoogleApi> endUserGoogleApiProvider;
+  private final FirecloudApiClientFactory firecloudApiClientFactory;
+  private final SamApiClientFactory samApiClientFactory;
   private final SamRetryHandler samRetryHandler;
+  private final Provider<GoogleApi> endUserGoogleApiProvider;
 
   @Autowired
   public IamServiceImpl(
-      SamApiClientFactory samApiClientFactory,
       CloudIamClient cloudIamClient,
       CloudResourceManagerService cloudResourceManagerService,
-      @Qualifier(SAM_END_USER_GOOGLE_API) Provider<GoogleApi> endUserGoogleApiProvider,
-      SamRetryHandler samRetryHandler) {
-    this.samApiClientFactory = samApiClientFactory;
+      FirecloudApiClientFactory firecloudApiClientFactory,
+      SamApiClientFactory samApiClientFactory,
+      SamRetryHandler samRetryHandler,
+      @Qualifier(SAM_END_USER_GOOGLE_API) Provider<GoogleApi> endUserGoogleApiProvider) {
     this.cloudIamClient = cloudIamClient;
     this.cloudResourceManagerService = cloudResourceManagerService;
-    this.endUserGoogleApiProvider = endUserGoogleApiProvider;
+    this.firecloudApiClientFactory = firecloudApiClientFactory;
+    this.samApiClientFactory = samApiClientFactory;
     this.samRetryHandler = samRetryHandler;
+    this.endUserGoogleApiProvider = endUserGoogleApiProvider;
   }
 
-  @Override
-  public void grantWorkflowRunnerRoleToCurrentUser(String googleProject) {
-    String petServiceAccountName =
-        getOrCreatePetServiceAccount(googleProject, endUserGoogleApiProvider.get());
-    grantServiceAccountUserRole(googleProject, petServiceAccountName);
-    grantLifeScienceRunnerRole(googleProject, Collections.singletonList(petServiceAccountName));
+  /** Gets a Terra pet service account from SAM. SAM will create one if one does not yet exist. */
+  private String getOrCreatePetServiceAccount(String googleProject, GoogleApi googleApi) {
+    return samRetryHandler.run((context) -> googleApi.getPetServiceAccount(googleProject));
   }
 
-  /** Gets a Terra pet service account from SAM. SAM will create one if user does not have it. */
-  private String getOrCreatePetServiceAccount(String googleProject, GoogleApi googleApi) {
-    return samRetryHandler.run(
-        (context) -> {
-          return googleApi.getPetServiceAccount(googleProject);
-        });
+  /**
+   * Gets a Terra pet service account from SAM as the current user. SAM will create one if one does
+   * not yet exist.
+   */
+  private String getOrCreatePetServiceAccountAsCurrentUser(String googleProject) {
+    return getOrCreatePetServiceAccount(googleProject, endUserGoogleApiProvider.get());
+  }
+
+  /**
+   * Gets a Terra pet service account from SAM as the given user using impersonation, if possible.
+   *
+   * <p>If the user has not yet accepted the latest Terms of Service, impersonation will not be
```

Comment:
Inspired by conversations in https://docs.google.com/document/d/1tVA-Pv6R65N2fJQw74-StgEmYAKkZLjj5WyeE9R-DGs/edit#